Refactor haversine_distance to a header-only API (#477)
rapidsai/cuspatial: cpp/doc/libcuspatial_refactoring_guide.md

The good news is that by maintaining the existing libcudf-based C++ API as a layer above the header-
only libcuspatial API, we can avoid problem 1 and problem 2 for users of the legacy API.

## Example API

Following is an example iterator-based API for `cuspatial::haversine_distance`.

```C++
template <class LonLatItA,
          class LonLatItB,
          class OutputIt,
          class Location = typename std::iterator_traits<LonLatItA>::value_type,
          class T        = typename Location::value_type>
OutputIt haversine_distance(LonLatItA a_lonlat_first,
                            LonLatItA a_lonlat_last,
                            LonLatItB b_lonlat_first,
                            OutputIt distance_first,
                            T const radius = EARTH_RADIUS_KM,
                            rmm::cuda_stream_view stream = rmm::cuda_stream_default);
```

There are a few key points to notice.

  1. The API is very similar to STL algorithms such as `std::transform`.
  2. All array inputs and outputs are iterator type templates. 
  3. Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`

>      type (include/cuspatial/types.hpp)

Reviewer: The code doesn't currently tell me this. Maybe add a `static_assert` to demonstrate it.

Reviewer: Concepts would be so nice for this :(

Author: I don't understand, I already have that static assert. 

```
static_assert(
    std::conjunction_v<std::is_same<lonlat_2d<T>, Location>, std::is_same<lonlat_2d<T>, LocationB>>,
    "Inputs must be cuspatial::lonlat_2d");
```

Author: Mentioned the static asserts in the refactoring guide.

Reviewer: Reading the code on lines33-43 and then reading on it describes the important things about this code...

"There are a few key points to notice...

Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d` type"

My only point is that it's impossible to "notice" this point just from the code as is on 33-43.

Author: Got it. I added this clarification to that point: 

```
 Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`
     type (include/cuspatial/types.hpp). This is enforced using a `static_assert` in the function
     body (discussed later).
```

  7. The size of the input and output ranges in the example API are equal, so the start and end of
     only the A range is provided (`a_lonlat_first` and `a_lonlat_last`). This mirrors STL APIs.
  8. This API returns an iterator to the element past the last element written to the output. This
     is inspired by `std::transform`, even though as with `transform`, many uses of 
     `haversine_distance` will not need this returned iterator.
  9. All APIs that run CUDA device code (including Thrust algorithms) or allocate memory take a CUDA
     stream on which to execute the device code and allocate memory.

## Example Documentation

Following is the (Doxygen) documentation for the above `cuspatial::haversine_distance`.

```C++
/**
 * @brief Compute haversine distances between points in set A to the corresponding points in set B.
 *
 * Computes N haversine distances, where N is `std::distance(a_lon_first, a_lon_last)`.
 * The distance for each `(a_lon[i], a_lat[i])` and `(b_lon[i], b_lat[i])` point pair is assigned to
 * `distance_first[i]`. `distance_first` must be an iterator to output storage allocated for N

>  * distances.

Reviewer: I find it nice when documentation explicitly calls out if it's okay for any of the iterator ranges to overlap. On the other hand, you could argue that unless it explicitly says that they cannot then it implies that they can.

E.g., `thrust::scatter` references this in the "Precondition" section: https://thrust.github.io/doc/group__scattering_gac1b33a02dd67cf9c8b3b3aa05c881a37.html

(You may want to consider having an explicit "Precondition" section. I've been leaning more and more in this direction for documentation.)

Author: Done. Used @pre and added preconditions regarding overlapping ranges, modeled after the Thrust docs.

 *
 * https://en.wikipedia.org/wiki/Haversine_formula
 *
 * @param[in]  a_lonlat_first: beginning of range of (longitude, latitude) locations in set A
 * @param[in]  a_lonlat_last: end of range of (longitude, latitude) locations in set A
 * @param[in]  b_lonlat_first: beginning of range of (longitude, latitude) locations in set B
 * @param[out] distance_first: beginning of output range of haversine distances
 * @param[in]  radius: radius of the sphere on which the points reside. default: 6371.0
 *            (approximate radius of Earth in km)
 * @param[in]  stream: The CUDA stream on which to perform computations and allocate memory.
 *

>  * All iterators must have the same floating-point `value_type`.

Reviewer: This could be potentially confusing given that the `lonlat` iterators are expected to have a custom struct `value_type`.

Author: It's actually the `value_type` of the `value_type`. i.e you can't pass one iterator to `lonlat_2d<float>` and one to `lonlat_2d<double>`.

Author: Clarified.

# cuSpatial C++ API Refactoring Guide

The original cuSpatial C++ API (libcuspatial) was designed to depend on RAPIDS libcudf and use
its core data types, especially `cudf::column`. For users who do not also use libcudf or other 
RAPIDS APIS, depending on libcudf could be a big barrier to adoption of libcuspatial. libcudf is 
a very large library and building it takes a lot of time.

Therefore, we are developing a standalone libcuspatial C++ API that does not depend on libcudf. This 
is a header-only template API with an iterator-based interface. This has a number of advantages

  1. With a header-only API, users can include and build exactly what they use.
  2. With a templated API, the API can be flexible to support a variety of basic data types, such 
     as float and double for positional data, and different integer sizes for indices.
  3. By templating on iterator types, cuSpatial algorithms can be fused with transformations of the
     input data, by using "fancy" iterators. Examples include transform iterators and counting
     iterators.
  4. Memory resources only need to be part of APIs that allocate temporary intermediate storage.

>      Output storage need not be allocated inside the API using a memory resource, since

Reviewer 2: ```suggestion
     Output storage is allocated outside of the API and only the output iterator is passed. 
```

     type (include/cuspatial/types.hpp)
  4. The `Location` type is a template that is by default equal to the `value_type` of the input
     iterators.
  5. The floating point type is a template (`T`) that is by default equal to the `value_type` of
     `Location`.
  6. The iterator types for the two input ranges (A and B) are distinct templates. This is crucial
     to enable composition of fancy iterators that may be different types for A and B.
  7. The size of the input and output ranges in the example API are equal, so the start and end of
     only the A range is provided (`a_lonlat_first` and `a_lonlat_last`). This mirrors STL APIs.
  8. This API returns an iterator to the element past the last element written to the output. This
     is inspired by `std::transform`, even though as with `transform`, many uses of 
     `haversine_distance` will not need this returned iterator.
  9. All APIs that run CUDA device code (including Thrust algorithms) or allocate memory take a CUDA
     stream on which to execute the device code and allocate memory.

## Example Documentation

Following is the (Doxygen) documentation for the above `cuspatial::haversine_distance`.

```C++
/**

>  * brief Compute haversine distances between points in set A to the corresponding points in set B.

Reviewer 2: ```suggestion
 * @brief Compute haversine distances between points in set A to the corresponding points in set B.
```

 * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
 * @tparam OutputIt Output iterator. Must meet the requirements of
 * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
 * @tparam Location The `value_type` of `LonLatItA` and `LonLatItB`. Must be
 * `cuspatial::lonlat_2d<T>`.
 * @tparam T The underlying coordinate type. Must be a floating-point type.
 *
 * @return Output iterator to the element past the last distance computed.
 *
 * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
 * "LegacyRandomAccessIterator"
 */
```

Key points:

  1. Precisely and succinctly documents what the API computes, and provides references.
  2. All parameters and all template parameters are documented.
  3. States the C++ standard iterator concepts that must be implemented, and that iterators must be
     device-accessible.
  4. Documents requirements such as that the input iterators must have `value_type` of
     `cuspatial::lonlat_2d<T>`, and that all iterators must have the same floating-point 
     `value_type`.
  5. Documents the units of any inputs or outputs that have them.

## cuSpatial libcudf-based C++ API (legacy API)

> [Note: I hesitate to call this the legacy API because we have not decided to deprecate it yet.]

Reviewer 2: That's fine, but should we keep this note?

is a header-only template API with an iterator-based interface. This has a number of advantages

  1. With a header-only API, users can include and build exactly what they use.
  2. With a templated API, the API can be flexible to support a variety of basic data types, such 
     as float and double for positional data, and different integer sizes for indices.
  3. By templating on iterator types, cuSpatial algorithms can be fused with transformations of the
     input data, by using "fancy" iterators. Examples include transform iterators and counting
     iterators.
  4. Memory resources only need to be part of APIs that allocate temporary intermediate storage.
     Output storage need not be allocated inside the API using a memory resource, since

The main disadvantages of this type of API are

  1. Header-only APIs can increase compilation time for code that depends on them.
  2. Some users (especially our Python API) may prefer a cuDF-based API.

The good news is that by maintaining the existing libcudf-based C++ API as a layer above the header-
only libcuspatial API, we can avoid problem 1 and problem 2 for users of the legacy API.

## Example API

> Following is an example iterator-based API for `cuspatial::haversine_distance`.

Reviewer: Totally tangential, but `cuspatial::` is right on the edge of being too long for a namespace imo. 

It's too bad [`cusp::`](https://github.com/cusplibrary/cusplibrary) is taken by a sparse linear algebra library that also happens to build on Thrust.

 * "LegacyRandomAccessIterator"
 */
```

Key points:

  1. Precisely and succinctly documents what the API computes, and provides references.
  2. All parameters and all template parameters are documented.
  3. States the C++ standard iterator concepts that must be implemented, and that iterators must be
     device-accessible.
  4. Documents requirements such as that the input iterators must have `value_type` of
     `cuspatial::location_2d<T>`, and that all iterators must have the same floating-point 
     `value_type`.
  5. Documents the units of any inputs or outputs that have them.

## cuSpatial libcudf-based C++ API (legacy API)

[Note: I hesitate to call this the legacy API because we have not decided to deprecate it yet.]

This is the existing API, unchanged by refactoring. Here is the existing 
`cuspatial::haversine_distance`:

```C++
std::unique_ptr<cudf::column> haversine_distance(
  cudf::column_view const& a_lon,
  cudf::column_view const& a_lat,
  cudf::column_view const& b_lon,
  cudf::column_view const& b_lat,
  double const radius                 = EARTH_RADIUS_KM,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

key points:

>   1. All input data are `cudf::column_view`. This is a type0erased container so determining the 

Author: Fixed ("type-erased")

a simple matter of a few static asserts and dynamic expectation checks, followed by a call to
`thrust::transform` with a custom transform functor.

```C++
template <class LonLatItA,
          class LonLatItB,
          class OutputIt,
          class Location = typename std::iterator_traits<LonLatItA>::value_type,
          class T        = typename Location::value_type>
OutputIt haversine_distance(LonLatItA a_lonlat_first,
                            LonLatItA a_lonlat_last,
                            LonLatItB b_lonlat_first,
                            OutputIt distance_first,
                            T const radius,
                            rmm::cuda_stream_view stream)
{
  using LocationB = typename std::iterator_traits<LonLatItB>::value_type;
  static_assert(std::conjunction_v<std::is_same<location_2d<T>, Location>,
                                   std::is_same<location_2d<T>, LocationB>>,
                "Inputs must be cuspatial::location_2d");
  static_assert(
    std::conjunction_v<std::is_floating_point<T>,
                       std::is_floating_point<typename LocationB::value_type>,

>                        std::is_floating_point<typename std::iterator_traits<OutputIt>::value_type>>,

Reviewer 2: Should we also assert that the precision of these types align?

Author: Do you mean that the `T` in each of the `Location` types is the same (`float`+`float`, `double`+`double` etc.?)  The `static_assert` just before this one ensures this is the case. 

```
 static_assert(
    std::conjunction_v<std::is_same<lonlat_2d<T>, Location>, std::is_same<lonlat_2d<T>, LocationB>>,
    "Inputs must be cuspatial::lonlat_2d");
```

If either is no `lonlat_2d<T>` then it will fail.
